dashboard/test/controllers/registrations_controller_test.rb

```diff
@@ -21,6 +21,127 @@ class RegistrationsControllerTest < ActionController::TestCase
     assert_response :success
   end
 
+  test "update: returns bad_request if user param is nil" do
+    student = create(:student)
+    sign_in student
+
+    put :update, params: {}
+    assert_response :bad_request
+  end
+
+  test "update: returns bad_request if user_type param is present" do
+    student = create(:student)
+    sign_in student
+
+    put :update, params: {user: {user_type: 'student'}}
+    assert_response :bad_request
+  end
+
+  test "update: returns bad_request if email param is present" do
+    student = create(:student)
+    sign_in student
+
+    put :update, params: {user: {email: 'example@email.com'}}
+    assert_response :bad_request
+  end
+
+  test "update: returns bad_request if hashed_email param is present" do
+    student = create(:student)
+    sign_in student
+
+    put :update, params: {user: {hashed_email: 'abcdef'}}
+    assert_response :bad_request
+  end
+
+  test "update: does not update user password if user cannot edit password" do
+    teacher = create(:teacher, :with_migrated_email_authentication_option, password: 'mypassword')
+    sign_in teacher
+
+    User.any_instance.stubs(:can_edit_password?).returns(false)
+
+    put :update, params: {user: {current_password: 'notmypassword', password: 'newpassword', password_confirmation: 'newpassword'}}
+    teacher.reload
+    assert_response :success
```

Comment:
yeah, i was confused until i realized that it returns a success response because it re-renders the edit view since the save failed. so the response is"success," but only because the edit template is re-rendered and the changes weren't saved to the db